### ISSUE TITLE
[Troubleshoot]: Unstable env/success_once when training pi05 on RoboCasa

### ISSUE THREAD

**Issue by pp489** [RLinf/RLinf#1044]
### Problem description

Hi, thanks for your great work on this project!

I’m currently training pi05 on RoboCasa using RL, and I encountered an issue regarding training stability. I would really appreciate any insights. 
During training, the metric `env/success_once` keeps fluctuating and does not show a clear upward trend.
- It oscillates roughly between 0.5 and 0.6
- No upward trend is observed even after 20 training steps

Setup
- Base model: I'm using a pi05 checkpoint that I fine-tuned myself on RoboCasa (not the official pretrained weights), and then applying RL on top of it.
- Algorithm: PPO (RL fine-tuning)

Setup
  Base model: I'm using a pi05 checkpoint that I fine-tuned myself on RoboCasa (not the official pretrained weights), and then applying RL on top of it.
  Environment: RoboCasa
  Algorithm: PPO (RL fine-tuning)

Additionally, I would greatly appreciate it if you could share any reference configurations or successful training curves for pi05 on RoboCasa, if available. I can provide more detailed TensorBoard logs or specific metrics if there’s anything you’d like me to check.

### Configuration YAML file

YAML config file:

You can also paste the full config here.
```
defaults:
  - env/robocasa@env.train
  - env/robocasa@env.eval
  - model/pi0@actor.model
  - training_backend/fsdp@actor.fsdp_config
  - override hydra/job_logging: stdout

hydra:
  run:
    dir: .
  output_subdir: null
  searchpath:
    - file://${oc.env:EMBODIED_PATH}/config/

cluster:
  num_nodes: 1
  component_placement:
    env,rollout,actor: 1,3,4,5

runner:
  task_type: embodied
  logger:
    log_path: "../results"
    project_name: rlinf
    experiment_name: "robocasa_ppo_openpi"
    logger_backends: ["tensorboard"] # wandb, swanlab

  max_epochs: 1000
  max_steps: -1
  only_eval: False
  val_check_interval: 5 #-1
  save_interval: 10 #50

  resume_dir: null #"/mnt/piaohongjing/RLinf/logs_2/20260413-17:45:21-test_update_ppo/robocasa_ppo_openpi/checkpoints/global_step_10" # Optional: path to a saved checkpoint directory, such as 'checkpoints/global_step_10'. If not None, it will be used to resume training.
  ckpt_path: null  # Optional: path to a .pt checkpoint. If not None, it will be loaded after the model is instantiated (for evaluation).

algorithm:
  normalize_advantages: True
  kl_penalty: kl  # how to estimate kl divergence: kl or kl_penalty
  group_size: 1 #8
  reward_coef: 1.0

  rollout_epoch: 16 #8
  eval_rollout_epoch: 4 #2  # set eval_rollout_epoch > 0 when enable runner.only_eval or runner.val_check_interval > 0
  reward_type: chunk_level
  logprob_type: action_level #chunk_level
  entropy_type: token_level

  update_epoch: 1 #4
  adv_type: gae
  loss_type: actor_critic
  loss_agg_func: "token-mean"
  kl_beta: 0.0 #last
  entropy_bonus: 0
  clip_ratio_high: 0.2
  clip_ratio_low: 0.2
  clip_ratio_c: 3.0
  value_clip: 0.2
  huber_delta: 10.0

  gamma: 0.99
  gae_lambda: 0.95 

  filter_rewards: False
  rewards_lower_bound: 0.1
  rewards_upper_bound: 0.9
  # params for generation
  sampling_params:
    do_sample: True
    temperature_train: 1.0
    temperature_eval: 0.6
    top_k: 50
    top_p: 1.0
    repetition_penalty: 1.0
    add_BOS: False

  # length argument for autoregressive sampling
  # max length means max amount of tokens to generate
  length_params:
    max_new_token: null
    max_length: 1024
    min_length: 1

env:
  group_name: "EnvGroup"
  # Use fixed seed across all workers to ensure same object types everywhere
  use_fixed_seed: True

  train:
    total_num_envs: 32 #64
    max_episode_steps: 500 # max episode steps for truncation
    max_steps_per_rollout_epoch: 500
    reward_coef: ${algorithm.reward_coef}
    group_size: ${algorithm.group_size}
    deterministic_reset: False
    seed: 0  # Fixed seed for object types
    #video_cfg:
    #  save_video: True
    #  info_on_video: True
    #  video_base_dir: ${runner.logger.log_path}/video/train
  eval:
    total_num_envs: 24 #8
    auto_reset: False #True
    ignore_terminations: False #True
    max_episode_steps: 500 #300
    max_steps_per_rollout_epoch: 500 #300
    group_size: 1
    is_eval: True 
    deterministic_reset: False
    seed: 0  # Same seed as train
    use_fixed_seed: True  # Use same seed across all workers
    video_cfg:
      save_video: True
      info_on_video: True
      video_base_dir: ${runner.logger.log_path}/video/eval

rollout:
  group_name: "RolloutGroup"
  generation_backend: "huggingface"
  recompute_logprobs: False
  unnorm_key: train/robocasa #robocasa
  enable_offload: True #False
  pipeline_stage_num: 1

  model:
    model_path: "/mnt/model/my_pi05_robocasa_pretrain_torch"  # Update this to your pi0 model path
    precision: ${actor.model.precision}

actor:
  group_name: "ActorGroup"
  training_backend: "fsdp"
  micro_batch_size: 32 #128
  global_batch_size: 2560 #2048  # 960 #1920 #
  seed: 42
  enable_offload: True #False

  # Override the default values in model/pi0
  model:
    model_path: "/mnt/model/my_pi05_robocasa_pretrain_torch"  # Update this to your pi0 model path
    num_action_chunks: 50
    action_dim: 32 #12  # Pi0 outputs 32D, extract first 12D, then convert [5:12] to PandaOmron's 12D format
    num_steps: 10 #5
    add_value_head: True
    # openpi specific parameters
    openpi:
      config_name: "my_pi05_robocasa_pretrain"
      num_images_in_input: 2  # base image + wrist image
      num_steps: ${actor.model.num_steps}
      pi05: True
      detach_critic_input: True

  optim:
    lr: 1.0e-5 #last
    value_lr: 3.0e-4 #1.0e-4
    adam_beta1: 0.9
    adam_beta2: 0.95
    adam_eps: 1.0e-08
    weight_decay: 0.01
    clip_grad: 2.0 #last
    critic_warmup_steps: 6 #0

  # Override the default values in training_backend/fsdp
  fsdp_config:
    strategy: "fsdp"
    sharding_strategy: "full_shard" #"no_shard"
    gradient_checkpointing: False # for openpi, gradient checkpointing is not supported, please do not change this value
    mixed_precision:
      param_dtype: ${actor.model.precision}
      reduce_dtype: ${actor.model.precision}
      buffer_dtype: ${actor.model.precision}

reward:
  use_reward_model: False

critic:
  use_critic_model: False

```

### Log file

You can find the log file in `logs/` folder or the $output_dir/$experiment_name folder (defined in the yaml config) if you are using our example scripts.

Log file:

If you cannot find the log, please provide the full log messages here.
```

[run_embodiment.log.gz](https://github.com/user-attachments/files/26744185/run_embodiment.log.gz)

<paste log here>
```

### Environment

python -V: Python 3.11.14
uv pip list:
Package                            Version     Editable project location
---------------------------------- ----------- --------------------------------------
absl-py                            2.4.0
accelerate                         1.12.0
aiohappyeyeballs                   2.6.1
aiohttp                            3.13.3
aiohttp-cors                       0.8.1
aiosignal                          1.4.0
annotated-types                    0.7.0
anyio                              4.13.0
argcomplete                        3.6.3
array-record                       0.8.3
astunparse                         1.6.3
attrs                              25.4.0
augmax                             0.4.1
av                                 17.0.0
bddl                               3.6.0
beartype                           0.19.0
beautifulsoup4                     4.14.3
blinker                            1.9.0
boto                               2.49.0
cachetools                         5.5.2
certifi                            2026.1.4
cffi                               2.0.0
cfgv                               3.5.0
charset-normalizer                 3.4.4
chex                               0.1.90
click                              8.3.1
cloudpickle                        3.1.2
cmake                              4.3.1
colorful                           0.6.0a1
contourpy                          1.3.3
crcmod                             1.7
cryptography                       43.0.3
cycler                             0.12.1
daqp                               0.8.5
datasets                           3.6.0
debugpy                            1.8.20
decorator                          5.2.1
deepdiff                           9.0.0
deprecated                         1.3.1
diffusers                          0.37.1
dill                               0.3.8
distlib                            0.4.0
dm-control                         1.0.38
dm-env                             1.6
dm-tree                            0.1.10
docstring-parser                   0.17.0
donfig                             0.8.1.post1
draccus                            0.10.0
easydict                           1.13
einops                             0.8.2
equinox                            0.13.6
etils                              1.14.0
evdev                              1.9.3
farama-notifications               0.0.4
fasteners                          0.20
fastjsonschema                     2.21.2
filelock                           3.20.3
flash-attn                         2.7.4.post1
flask                              3.1.3
flatbuffers                        25.12.19
flax                               0.10.2
fonttools                          4.62.1
frozenlist                         1.8.0
fsspec                             2025.3.0
future                             1.0.0
gast                               0.7.0
gcs-oauth2-boto-plugin             3.3
gcsfs                              2025.3.0
gdown                              5.2.1
gitdb                              4.0.12
gitpython                          3.1.46
glfw                               2.10.0
google-api-core                    2.29.0
google-apitools                    0.5.35
google-auth                        2.39.0
google-auth-httplib2               0.3.0
google-auth-oauthlib               1.3.1
google-cloud-core                  2.5.1
google-cloud-storage               3.10.1
google-crc32c                      1.8.0
google-pasta                       0.2.0
google-reauth                      0.1.1
google-resumable-media             2.8.2
googleapis-common-protos           1.72.0
grpcio                             1.78.0
gsutil                             5.35
gym                                0.26.2
gym-aloha                          0.1.3
gym-notices                        0.1.0
gymnasium                          0.29.1
h11                                0.16.0
h5py                               3.14.0
hf-transfer                        0.1.9
hf-xet                             1.3.0b0
hidapi                             0.15.0
httpcore                           1.0.9
httplib2                           0.20.4
httpx                              0.28.1
huggingface-hub                    0.36.2
humanize                           4.15.0
hydra-core                         1.4.0.dev1
icmplib                            3.0.4
identify                           2.6.16
idna                               3.11
imageio                            2.37.2
imageio-ffmpeg                     0.6.0
immutabledict                      4.3.1
importlib-metadata                 8.0.0
iniconfig                          2.3.0
inquirerpy                         0.3.4
itsdangerous                       2.2.0
jax                                0.5.3
jax-cuda12-pjrt                    0.5.3
jax-cuda12-plugin                  0.5.3
jaxlib                             0.5.3
jaxtyping                          0.2.36
jinja2                             3.1.6
joblib                             1.5.3
jsonlines                          4.0.0
jsonschema                         4.26.0
jsonschema-specifications          2025.9.1
jupyter-core                       5.9.1
jupytext                           1.19.1
keras                              3.13.2
kiwisolver                         1.5.0
labmaze                            1.0.6
lerobot                            0.1.0
libclang                           18.1.1
liger-kernel                       0.6.5
llvmlite                           0.44.0
lxml                               6.0.2
markdown                           3.10.2
markdown-it-py                     4.0.0
markupsafe                         3.0.3
matplotlib                         3.10.8
mdit-py-plugins                    0.5.0
mdurl                              0.1.2
mergedeep                          1.3.4
mink                               0.0.13
ml-collections                     1.0.0
ml-dtypes                          0.5.4
monotonic                          1.6
mpmath                             1.3.0
msgpack                            1.1.2
mujoco                             3.2.6
multidict                          6.7.1
multiprocess                       0.70.16
mypy-extensions                    1.1.0
namex                              0.1.0
nbformat                           5.10.4
nest-asyncio                       1.6.0
networkx                           3.6.1
ninja                              1.13.0
nltk                               3.9.4
nodeenv                            1.10.0
numba                              0.61.2
numcodecs                          0.16.5
numpy                              1.26.4
numpydantic                        1.8.0
nvidia-cublas-cu12                 12.4.5.8
nvidia-cuda-cupti-cu12             12.4.127
nvidia-cuda-nvcc-cu12              12.9.86
nvidia-cuda-nvrtc-cu12             12.4.127
nvidia-cuda-runtime-cu12           12.4.127
nvidia-cudnn-cu12                  9.1.0.70
nvidia-cufft-cu12                  11.2.1.3
nvidia-curand-cu12                 10.3.5.147
nvidia-cusolver-cu12               11.6.1.9
nvidia-cusparse-cu12               12.3.1.170
nvidia-cusparselt-cu12             0.6.2
nvidia-ml-py                       13.590.48
nvidia-nccl-cu12                   2.21.5
nvidia-nvjitlink-cu12              12.4.127
nvidia-nvtx-cu12                   12.4.127
nvitop                             1.6.2
oauth2client                       4.1.3
oauthlib                           3.3.1
omegaconf                          2.4.0.dev4
opencensus                         0.11.4
opencensus-context                 0.2.dev0
opencv-python                      4.11.0.86
opencv-python-headless             4.11.0.86
openexr                            3.4.8
openpi                             0.1.0
openpi-client                      0.1.0
opentelemetry-api                  1.26.0
opentelemetry-exporter-prometheus  0.47b0
opentelemetry-proto                1.39.1
opentelemetry-sdk                  1.26.0
opentelemetry-semantic-conventions 0.47b0
opt-einsum                         3.4.0
optax                              0.2.8
optree                             0.19.0
orbax-checkpoint                   0.11.13
orderly-set                        5.5.0
packaging                          26.0
pandas                             3.0.0
peft                               0.11.1
pfzy                               0.3.4
pillow                             12.1.1
pip                                26.0.1
platformdirs                       4.5.1
pluggy                             1.6.0
polars                             1.39.3
polars-runtime-32                  1.39.3
pre-commit                         4.5.1
prettytable                        3.17.0
prometheus-client                  0.24.1
promise                            2.3
prompt-toolkit                     3.0.52
propcache                          0.4.1
proto-plus                         1.27.1
protobuf                           6.33.0
psutil                             7.2.2
py-spy                             0.4.1
pyarrow                            23.0.0
pyasn1                             0.6.2
pyasn1-modules                     0.4.2
pybind11                           3.0.1
pycparser                          3.0
pydantic                           2.12.5
pydantic-core                      2.41.5
pyecharts                          2.1.0
pygame                             2.6.1
pygments                           2.19.2
pymunk                             7.2.0
pynput                             1.8.1
pyopengl                           3.1.10
pyopenssl                          24.2.1
pyparsing                          3.3.2
pyperclip                          1.11.0
pyrealsense2                       2.56.5.9235
pysocks                            1.7.1
pytest                             9.0.2
python-dateutil                    2.9.0.post0
python-xlib                        0.33
pyu2f                              0.1.5
pyyaml                             6.0.3
pyyaml-include                     1.4.1
pyzmq                              27.1.0
qpsolvers                          4.11.0
ray                                2.47.1
referencing                        0.37.0
regex                              2026.1.15
requests                           2.32.5
requests-oauthlib                  2.0.0
rerun-sdk                          0.23.1
retry-decorator                    2.0a1
rich                               14.3.3
robocasa                           0.2.0       /mnt/piaohongjing/RLinf/.venv/robocasa
robosuite                          1.5.1
rpds-py                            0.30.0
rsa                                4.7.2
ruff                               0.15.0
safetensors                        0.7.0
sapien                             3.0.1
scipy                              1.17.0
sentencepiece                      0.2.1
sentry-sdk                         3.0.0a7
setuptools                         75.8.2
simple-parsing                     0.1.8
simplejson                         3.20.2
six                                1.17.0
smart-open                         7.5.0
smmap                              5.0.2
soupsieve                          2.8.3
svgwrite                           1.4.3
swanlab                            0.7.8
sympy                              1.13.1
tensorboard                        2.20.0
tensorboard-data-server            0.7.2
tensorflow                         2.21.0
tensorflow-addons                  0.23.0
tensorflow-datasets                4.9.9
tensorflow-graphics                2021.12.3
tensorflow-metadata                1.17.3
tensorstore                        0.1.82
termcolor                          3.3.0
tianshou                           0.4.10
timm                               0.9.10
tokenizers                         0.21.4
toml                               0.10.2
toolz                              1.1.0
torch                              2.6.0
torch-tb-profiler                  0.4.3
torchcodec                         0.2.0
torchdata                          0.11.0
torchvision                        0.21.0
tqdm                               4.67.3
tqdm-loggable                      0.4.1
traitlets                          5.14.3
transformers                       4.53.2
transforms3d                       0.4.2
tree                               0.2.4
treescope                          0.1.10
trimesh                            4.11.5
triton                             3.2.0
typeguard                          4.5.1
typing-extensions                  4.15.0
typing-inspect                     0.9.0
typing-inspection                  0.4.2
tyro                               1.0.11
urllib3                            2.6.3
uv                                 0.10.2
virtualenv                         20.36.1
wadler-lindig                      0.1.7
wandb                              0.24.2
wcwidth                            0.6.0
websockets                         16.0
werkzeug                           3.1.5
wheel                              0.46.3
wrapt                              2.1.1
xxhash                             3.6.0
yarl                               1.22.0
zarr                               3.1.5
zipp                               3.23.0

nvidia-smi: 
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.90.07              Driver Version: 550.90.07      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A100-SXM4-80GB          On  |   00000000:3D:00.0 Off |                    0 |
| N/A   35C    P0             69W /  400W |   23643MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+  
RLinf version: v0.1-rc1-190-g874d268 
(forked from upstream, commit 874d2681f8b9d05efb827d462379c96babdbbd1c)
Docker image tag: 

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by snakefood3232** [comment]
Your `env/success_once` metric oscillating between 0.5 and 0.6 indicates a training stability issue with PPO on RoboCasa. I'd handle this by first analyzing your reward shaping and hyperparameters, particularly the PPO clip range and learning rate schedule. I can have a PR with config adjustments and diagnostic scripts up in a day. I've previously debugged similar unstable RL training curves in embodied AI pipelines.

**Comment by snakefood3232** [comment]
Your `env/success_once` metric oscillating between 0.5 and 0.6 suggests a PPO convergence issue, likely in hyperparameters or reward shaping. I'd analyze your config and TensorBoard logs to adjust the clipping range or advantage normalization. I can propose a fix based on similar RL fine-tuning work I've done for robotic policies. Can have a PR with tuned parameters up in a day.
